### PR TITLE
Bugfix termostat user interface value validation

### LINKED ISSUE #9815: When writing invalid temperature display mode using Mac Chip Tool, getting a success response
When writing invalid temperature display mode (2) using Mac Chip Tool, the expected output is a failure response but I am getting success response. 

Testcase: 3.2.2. [TC-TSUIC-2.2] Primary functionality with server as DUT
Cluster: Thermostat User Configuration

shyam@shyams-MacBook-Pro chip-tool % ./out/debug/chip-tool thermostatuserinterfaceconfiguration write temperature-display-mode 2 1
[1631918659593] [14766:717617] CHIP: [CTL] Read local id 0x000000000001B669, remote id 0x8B6CC7E459241F2C
[1631918659594] [14766:717617] CHIP: [CTL] Generating NOC
[1631918659594] [14766:717617] CHIP: [CTL] Generating ICAC
[1631918659595] [14766:717617] CHIP: [DL] _Init
[1631918659595] [14766:717617] CHIP: [IN] UDP::Init bind&listen port=5541
[1631918659595] [14766:717617] CHIP: [IN] UDP::Init bind&listen port=5541
[1631918659595] [14766:717617] CHIP: [IN] TransportMgr initialized
[1631918659595] [14766:717617] CHIP: [DIS] Init fabric pairing table with server storage
[1631918659595] [14766:717617] CHIP: [ZCL] Using ZAP configuration...
[1631918659595] [14766:717617] CHIP: [ZCL] deactivate report event
[1631918659596] [14766:717617] CHIP: [DIS] Verifying the received credentials
[1631918659596] [14766:717617] CHIP: [DIS] Added new fabric at index: 1, Initialized: 1
[1631918659596] [14766:717617] CHIP: [DIS] Assigned fabric ID: 0x0000000000000000, node ID: 0x000000000001B669
[1631918659596] [14766:717617] CHIP: [CTL] Joined the fabric at index 1
[1631918659597] [14766:717622] CHIP: [SC] Sent SigmaR1 msg
[1631918659614] [14766:717624] CHIP: [EM] Received message of type 0x31 with vendorId 0x0000 and protocolId 0x0000 on exchange 44249
[1631918659614] [14766:717624] CHIP: [EM] Rxd Ack; Removing MsgId:00000000 from Retrans Table
[1631918659614] [14766:717624] CHIP: [EM] Removed CHIP MsgId:00000000 from RetransTable
[1631918659614] [14766:717624] CHIP: [SC] Received SigmaR2 msg
[1631918659614] [14766:717624] CHIP: [SC] Peer assigned session key ID 15
[1631918659615] [14766:717624] CHIP: [SC] Sending SigmaR3
[1631918659615] [14766:717624] CHIP: [EM] Piggybacking Ack for MsgId:00000000 with msg
[1631918659615] [14766:717624] CHIP: [SC] Sent SigmaR3 msg
[1631918659615] [14766:717624] CHIP: [IN] New secure session created for device 0x8B6CC7E459241F2C, key 15!!
[1631918659615] [14766:717624] CHIP: [TOO] Sending cluster (0x0204) command (0x01) on endpoint 1
[1631918659615] [14766:717624] CHIP: [DMG] WriteClient moving to [Initialize]
[1631918659615] [14766:717624] CHIP: [DMG] WriteClient moving to [AddAttribu]
[1631918659615] [14766:717624] CHIP: [IN] Secure message was encrypted: Msg ID 0
[1631918659615] [14766:717624] CHIP: [IN] Encrypted message 0x7fe19d80e470 to 0x8B6CC7E459241F2C of type 6 and protocolId 5 on exchange 44250.
[1631918659615] [14766:717624] CHIP: [IN] Sending msg 0x7fe19d80e470 to 0x8B6CC7E459241F2C at utc time: 96883302 msec
[1631918659615] [14766:717624] CHIP: [IN] Sending secure msg on generic transport
[1631918659615] [14766:717624] CHIP: [IN] Secure msg send status ../../third_party/connectedhomeip/src/inet/IPEndPointBasis.cpp:915: Success
[1631918659616] [14766:717624] CHIP: [DMG] WriteClient moving to [AwaitingRe]
[1631918659626] [14766:717624] CHIP: [EM] Received message of type 0x10 with vendorId 0x0000 and protocolId 0x0000 on exchange 44249
[1631918659626] [14766:717624] CHIP: [EM] Rxd Ack; Removing MsgId:00000000 from Retrans Table
[1631918659626] [14766:717624] CHIP: [EM] Removed CHIP MsgId:00000000 from RetransTable
[1631918659626] [14766:717624] CHIP: [EM] Received message of type 0x07 with vendorId 0x0000 and protocolId 0x0005 on exchange 44250
[1631918659627] [14766:717624] CHIP: [EM] Rxd Ack; Removing MsgId:00000000 from Retrans Table
[1631918659627] [14766:717624] CHIP: [EM] Removed CHIP MsgId:00000000 from RetransTable
[1631918659627] [14766:717624] CHIP: [DMG] WriteResponse =
[1631918659627] [14766:717624] CHIP: [DMG] {
[1631918659627] [14766:717624] CHIP: [DMG] 	AttributeStatusList =
[1631918659627] [14766:717624] CHIP: [DMG] 	[
[1631918659627] [14766:717624] CHIP: [DMG] 		AttributeStatusElement =
[1631918659627] [14766:717624] CHIP: [DMG] 		{
[1631918659627] [14766:717624] CHIP: [DMG] 			AttributePath =
[1631918659627] [14766:717624] CHIP: [DMG] 			{
[1631918659627] [14766:717624] CHIP: [DMG] 				FieldTag = 0x0,
[1631918659627] [14766:717624] CHIP: [DMG] 				NodeId = 0x8b6cc7e459241f2c,
[1631918659627] [14766:717624] CHIP: [DMG] 				ClusterId = 0x204,
[1631918659627] [14766:717624] CHIP: [DMG] 				EndpointId = 0x1,
[1631918659627] [14766:717624] CHIP: [DMG] 			}
[1631918659627] [14766:717624] CHIP: [DMG] 				
[1631918659627] [14766:717624] CHIP: [DMG] 			StatusElement =
[1631918659627] [14766:717624] CHIP: [DMG] 			{
[1631918659627] [14766:717624] CHIP: [DMG] 				GeneralCode = 0x0,
[1631918659627] [14766:717624] CHIP: [DMG] 				ProtocolId = 0x0,
[1631918659627] [14766:717624] CHIP: [DMG] 				protocolCode = 0x0,
[1631918659627] [14766:717624] CHIP: [DMG] 			},
[1631918659627] [14766:717624] CHIP: [DMG] 			
[1631918659627] [14766:717624] CHIP: [DMG] 		},
[1631918659627] [14766:717624] CHIP: [DMG] 		
[1631918659627] [14766:717624] CHIP: [DMG] 	],
[1631918659627] [14766:717624] CHIP: [DMG] 	
[1631918659627] [14766:717624] CHIP: [DMG] }
[1631918659627] [14766:717624] CHIP: [ZCL] WriteResponse:
[1631918659627] [14766:717624] CHIP: [ZCL]   status: EMBER_ZCL_STATUS_SUCCESS (0x00)
[1631918659627] [14766:717624] CHIP: [TOO] Default Success Response
[1631918659627] [14766:717624] CHIP: [DMG] WriteClient moving to [Uninitiali]
[1631918659627] [14766:717624] CHIP: [EM] Sending Standalone Ack for MsgId:00000001
[1631918659627] [14766:717624] CHIP: [IN] Secure message was encrypted: Msg ID 1
[1631918659627] [14766:717624] CHIP: [IN] Encrypted message 0x70000f5bd518 to 0x8B6CC7E459241F2C of type 16 and protocolId 0 on exchange 44250.
[1631918659627] [14766:717624] CHIP: [IN] Sending msg 0x70000f5bd518 to 0x8B6CC7E459241F2C at utc time: 96883313 msec
[1631918659627] [14766:717624] CHIP: [IN] Sending secure msg on generic transport
[1631918659627] [14766:717624] CHIP: [IN] Secure msg send status ../../third_party/connectedhomeip/src/inet/IPEndPointBasis.cpp:915: Success
[1631918659627] [14766:717624] CHIP: [EM] Flushed pending ack for MsgId:00000001
[1631918659627] [14766:717617] CHIP: [CTL] Shutting down the commissioner
[1631918659627] [14766:717617] CHIP: [CTL] Shutting down the controller
[1631918659628] [14766:717617] CHIP: [DL] Inet Layer shutdown
[1631918659628] [14766:717617] CHIP: [DL] BLE layer shutdown
[1631918659628] [14766:717617] CHIP: [DL] System Layer shutdown

### PR DESCRIPTION
**Pull request by ghost**
#### Problem
* Fixes #9815

#### Change overview
* add checks

#### Testing
all-cluster-app and chip-device-ctrl.

### CODE REVIEW

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/all-clusters-app/linux/main.cpp [hunk 1]
```diff
@@ -42,6 +47,22 @@ Identify gIdentify1 = {
     EMBER_ZCL_IDENTIFY_IDENTIFY_TYPE_VISIBLE_LED,
 };
 
+Protocols::InteractionModel::Status MatterPreAttributeChangeCallback(const app::ConcreteAttributePath & attributePath, uint8_t mask,
+                                                                     uint8_t type, uint16_t size, uint8_t * value)
+{
+    Protocols::InteractionModel::Status status = Protocols::InteractionModel::Status::Success;
+    if (attributePath.mClusterId == app::Clusters::Thermostat::Id)
+    {
+        status = MatterThermostatClusterServerPreAttributeChangedCallback(attributePath, type, size, value);
```
[the remark this answers is not in the document]
Comment (chrisbeach59) [reply examples/all-clusters-app/linux/main.cpp:82]: As far as I know the cluster specific callbacks were not yet in place.

[the remark this answers is not in the document]
Comment (mrjerryjohns) [reply examples/all-clusters-app/linux/main.cpp:82]: I see the cluster-specific pre-change callbacks present in the tv-casting app:

```
    const EmberAfGenericClusterFunction chipFuncArrayIasZoneServer[] = {                                                           \
        (EmberAfGenericClusterFunction) emberAfIasZoneClusterServerInitCallback,                                                   \
        (EmberAfGenericClusterFunction) emberAfIasZoneClusterServerMessageSentCallback,                                            \
        (EmberAfGenericClusterFunction) MatterIasZoneClusterServerPreAttributeChangedCallback,                                     \
    };                                                                                                                             \
```

Not sure if it is working though. But if it isn't, we should endeavor to fix it.

[the remark this answers is not in the document]
Comment (vivien-apple) [reply examples/all-clusters-app/linux/main.cpp:82]: It should work, assuming the cluster has been added to: https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/templates/app/helper.js#L109

This is something I'm trying to get rid off though but it will takes time, so it sounds fair to reuse the linked mechanism.

[the remark this answers is not in the document]
Comment (eve-cxrp) [reply examples/all-clusters-app/linux/main.cpp]: @bzbarsky-apple removed the hook up through the `MatterPreAttributeChangeCallback`.

File: examples/all-clusters-app/linux/main.cpp [hunk 2]
```diff
@@ -16,12 +16,17 @@
...
+#include <app-common/zap-generated/callback.h>
+#include <app-common/zap-generated/ids/Clusters.h>
 #include <app/Command.h>
+#include <app/ConcreteAttributePath.h>
 #include <app/clusters/identify-server/identify-server.h>
 #include <app/util/af.h>
 
 #include "AppMain.h"
 
+using namespace chip;
```
Comment (bzbarsky-apple) [inline examples/all-clusters-app/linux/main.cpp:28]: Why are these changes needed?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
https://github.com/project-chip/connectedhomeip/pull/11587 to address my review comment.
